### PR TITLE
facilitator: multi-threaded aggregation

### PR DESCRIPTION
**Pull request by tgeoghegan**
We observed that in NCI's production deployment, us-ca-apple
individual aggregation tasks are often taking more than eight hours to
run, which means that even adding more aggregation workers won't speed
things up. We schedule aggregations to run every eight hours, making it
impossible for NCI to keep up with incoming work.

To ameliorate this, this commit refactors the aggregation task handler
so that it can spawn multiple threads and sum over individual batches in
parallel. The threads are managed using the `WorkQueue` struct, which
implements a no-producer, multiple-consumer work queue. I say
"no-producer" because it doesn't support an "enqueue" method as you
might see in most pub/sub implementations, and instead expects the
entire list of jobs to be available at the creation of the `WorkQueue`.
In our use case, we know up front the list of jobs we want to execute in
parallel, and this allows us to greatly simplify the queue
implementation. For instance, we can use a simple `Vec` to store jobs
instead of requiring a `VecDeque` or some more complex data structure.

To allow safely sharing or sending resources across worker threads, we
rely heavily on structs implementing the `Clone` trait. Mostly this is a
simple matter of tagging existing definitions with `#[derive(Clone)]`,
but it required some extra gymnastics for `S3Client` and its associated
structs:

	- `ClientProvider` was refactored from a supertrait over
	  `std::ops::Fn` into a more conventional trait to make it
	  easier to reason about how to `Clone` it.
	- Because Rusoto's `MockRequestDispatcher` & Co. are not
	  `Clone` and have private fields, we had to introduce some
	  proxy objects that can be cloned.

To enable multithreaded aggregation, pass `--thread-count=<int>` to
`facilitator aggregate-worker`, or set the `THREAD_COUNT` environment
variable to an integer value (e.g., in a Kubernetes ConfigMap).

### CODE REVIEW

**Comment by tgeoghegan** [comment]
This won't build as-is because it depends on https://github.com/abetterinternet/libprio-rs/pull/13.

File: facilitator/src/work_queue.rs [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    /// Send job results to the queue. Clients need not call this 1:1 with
+    /// dequeue_job, if they can reasonably represent there results of multiple
+    /// jobs in one R.
+    pub(crate) fn send_results(&mut self, result: R) {
+        self.results.lock().unwrap().push(result)
+    }
+
+    /// Get the results from this work queue's completed jobs, consuming the
+    /// work queue. It is the caller's responsibility to ensure that no workers
+    /// are still dequeueing jobs.
```
Comment (jsha) [inline facilitator/src/work_queue.rs:46]: This can enforce its expectations a little more strongly: It can check that `self.jobs` is empty, upon pain of `Err()`.

**Review by jsha — COMMENTED** [review]
Perhaps it's my Go leanings, but I would be somewhat tempted to use a [std::sync::mpsc::sync_channel](https://doc.rust-lang.org/stable/std/sync/mpsc/fn.sync_channel.html) to implement the results-collection side of `WorkQueue`.

File: facilitator/src/work_queue.rs (hunk 1 above)
Comment (tgeoghegan) [reply facilitator/src/work_queue.rs:46]: > Perhaps it's my Go leanings, but I would be somewhat tempted to use a [std::sync::mpsc::sync_channel](https://doc.rust-lang.org/stable/std/sync/mpsc/fn.sync_channel.html) to implement the results-collection side of `WorkQueue`.

That's a good idea! I had looked at `mpsc` as a replacement for all of `WorkQueue`, but realized that what I needed is `spmc`, which the stdlib doesn't provide. seanmonstar of Hyper fame has an [spmc crate](https://github.com/seanmonstar/spmc) but finally I decided against adding a dependency since `WorkQueue` is so simple. Anyway, I'll try using `std::sync::mpsc` for the results side and we'll see if it's any nicer.

Comment (tgeoghegan) [reply facilitator/src/work_queue.rs:46]: I gave this a try with `mpsc::SyncChannel` and found it to be. In my usage, I clone the `WorkQueue` into each worker thread so they can all send results. [`std::sync::mpsc::Receiver`](https://doc.rust-lang.org/stable/std/sync/mpsc/struct.Receiver.html) is not `Clone` because you're not meant to share it across threads. But what that means is that I can't have the send and receive ends of the queue abstracted behind a single object: the caller in `aggregation.rs` would have to hold on to both a `WorkQueue` (to be cloned off into worker threads) and a `std::sync::mpsc::Receiver` (to eventually get the results). Since using `mpsc::SyncChannel` doesn't make the implementation of `WorkQueue` any simpler, I'm going to stick with what I've got.

**Comment by tgeoghegan** [comment]
@jsha mentioned in Slack that I might want to convert the `callback` parameter that `Aggregate::generate_sum_part` gets into something more formal: a trait like `Supervisor` plus an implementation like `DeadlineExtendingSupervisor` for use in `facilitator.rs` and `CounterIncrementingSupervisor` for use in `integration_tests.rs`. The rationale is that it's easier to reason about ownership, cloneability, etc. of a struct than a closure. I agree, and at the risk of a bit of scope creep in this PR, I'm going to take on that change. While I'm doing that, I should be able to fix #319, too. But that will wait until Monday.

**Review by jsha — CHANGES_REQUESTED** [review]
Nice work!

File: facilitator/README.md [hunk 2]
```diff
@@ -44,6 +44,14 @@ AWS SNS/SQS support is experimental and has not been validated. To use it, pass
...
 
+## Configuration
+
+### `aggregate-worker` thread count
+
+The `aggregate` and `aggregate-worker` subcommands can be configured to spawn multiple worker threads that will sum over batches in an aggregation in parallel before reducing each batch's sum into a sum part that is transmitted to the portal server. By default, a single thread is used. Set either the `--thread-count` command line argument or the `THREAD_COUNT` environment variable (e.g., in a Kubernetes `ConfigMap`) to an integer value to spawn the corresponding number of threads.
+
+To make best use of a multithreaded `aggregate-worker` in Kubernetes, make sure to configure the CPU and memory limits appropriately so that multiple cores can be allocated. For instance, a CPU request of `0.1` and a CPU limit of `3` might work well for four worker threads, keeping in mind that since Prio batch processing involves a lot of network I/O, we would not expect four worker threads on four cores to see consistently high CPU utilization.
+
```
[the remark this answers is not in the document]
Comment (tgeoghegan) [reply facilitator/README.md:54]: > Since `aggregate` is not really used in prod, does it need to support parallelism?

No, but it's still best if `aggregate` and `aggregate-worker` are as similar as possible. Plus this way we don't have to maintain a serial implementation as well as the parallel one.

>In Go, parallelism (i.e. number of OS threads) is automatically determined by the number of CPUs reported by the machine at runtime startup. Does it make sense to do that here to reduce the amount of configuration synchronization people need?

In this case, I don't think one thread per CPU is a good idea. Especially when the storage is something like S3, `facilitator` is heavily I/O bound (i.e. we spend a lot of time waiting on responses from various cloud services), so threads will spend a lot of time parked in `read` or `write` and thus can allow other threads to execute.

File: facilitator/src/aggregation.rs [hunk 3]
```diff
@@ -109,10 +110,10 @@ impl<'a> BatchAggregator<'a> {
     /// Compute the sum part for all the provided batch IDs and write it out to
     /// the aggregation transport. The provided callback is invoked after each
     /// batch is aggregated.
-    pub fn generate_sum_part<F: FnMut() + Send + Clone + 'static>(
+    pub fn generate_sum_part<W: Watchdog>(
         &mut self,
         batches: Vec<SumInput>,
-        callback: F,
+        watchdog: &'a W,
```
[the remark this answers is not in the document]
Comment (tgeoghegan) [reply facilitator/src/aggregation.rs:116]: The difference is that `generate_sum_part` is multithreaded, while `BatchIntaker::generate_validation_share` is not. The latter calls `send_heartbeat()` directly on the `watchdog` it is passed while `generate_sum_part` uses `clone()` to safely send "copies" of the watchdog to each of the threads it spawns. the `'a` lifetime is a leftover from an earlier iteration where I was struggling to get the lifetimes right, and I've deleted it.

I agree that it's odd that the two method signatures don't match, and the reason why leaks the detail that one method is multithreaded and the other isn't. What do you think can be done to improve this? Take `&W: Watchdog` in both methods and always clone into a mutable instance, just for consistency?

File: facilitator/src/aggregation.rs [hunk 4]
```diff
@@ -94,61 +123,130 @@ impl<'a> BatchAggregator<'a> {
...
+                    .iter()
+                    .map(|k| Server::new(dimension, share_aggregator.is_first, k.clone()))
+                    .collect::<Vec<Server>>();
+                let mut individual_client_count = 0;
+                let mut invalid_packet_uuids = Vec::new();
+                loop {
+                    let job = match queue_clone.dequeue_job() {
+                        Some(job) => job,
+                        None => break,
+                    };
```
[the remark this answers is not in the document]
Comment (tgeoghegan) [reply facilitator/src/aggregation.rs:176]: Ooh, much better!

File: facilitator/src/task/sqs.rs [hunk 5]
```diff
@@ -53,6 +49,7 @@ impl<T: Task> TaskQueue<T> for AwsSqsTaskQueue<T> {
         );
 
         let client = self.sqs_client()?;
+        let runtime = basic_runtime()?;
```
[the remark this answers is not in the document]
Comment (tgeoghegan) [reply facilitator/src/task/sqs.rs:52]: [`tokio::runtime::Runtime`](https://docs.rs/tokio/1.3.0/tokio/runtime/struct.Runtime.html) does not implement `std::clone::Clone`, and so a struct which wants to derive `Clone` can't have a `Runtime` member. There is [`tokio::runtime::Handle`](https://tokio-rs.github.io/tokio/doc/tokio/runtime/struct.Handle.html) that can be cloned but I'm pretty sure we don't want that: `basic_runtime` uses [`tokio::runtime::Builder::new_current_thread`](https://docs.rs/tokio/1.3.0/tokio/runtime/struct.Builder.html#method.new_current_thread) so that we can synchronously run async tasks.

Suppose I start on the main thread of a process and then do:

```
let runtime = tokio::runtime::Builder::new_current_thread().build();
let handle_clone = runtime.handle().clone();
thread::spawn(move || {
    // now on thread B
    handle_clone.block_on(/* some work */);
});
```
The `handle_clone.block_on()` call will block thread B until it's finished, but what I'm not sure of is whether the `Future` passed to `block_on()` will execute on the main thread or thread B. That is, when you create a `new_current_thread` runtime, does that mean current thread at time of creation of the runtime, or current thread at the time of dispatching work onto the runtime?

In any case, one thing I could do better here is to provide a custom `std::clone::Clone()` impl on `AwsSqsTaskQueue` that would clone every other member of the struct but instantiate a new `Runtime`, so that then at least each instance of `AwsSqsTaskQueue` wouldn't create a new runtime every time it makes an API call.

[the remark this answers is not in the document]
Comment (tgeoghegan) [reply facilitator/src/task/sqs.rs:52]: I spoke too soon -- I can't instantiate a new `Runtime` in `std::clone::Clone::clone()` because `basic_runtime()` can fail, and `clone()` must always succeed. Also I ran an experiment with cloned `tokio::runtime::Handle`s and it's not promising:
```
#[test]
    fn test_current_thread() {
        print_current_thread_id("test function");

        let runtime = Builder::new_current_thread().enable_all().build().unwrap();
        runtime.block_on(async {
            print_current_thread_id("block_on runtime");
        });
        runtime.handle().spawn_blocking(|| {
            print_current_thread_id("handle spawn_blocking");
        });
        let handle = runtime.handle().clone();
        handle.spawn_blocking(|| {
            print_current_thread_id("handle clone spawn_blocking");
        });

        thread::spawn(move || {
            print_current_thread_id("thread");
            handle.spawn_blocking(|| {
                print_current_thread_id("spawn blocking from thread");
            });

            let runtime = Builder::new_current_thread().enable_all().build().unwrap();
            runtime.block_on(async {
                print_current_thread_id("block on from runtime created in thread");
            })
        }).join().unwrap();

        assert!(false);
    }

    fn print_current_thread_id(context: &str) {
        println!("current thread ID in {}: {:?}", context, current().id());
    }
```

yields:

```
current thread ID in test function: ThreadId(71)
current thread ID in block_on runtime: ThreadId(71)
current thread ID in handle spawn_blocking: ThreadId(72)
current thread ID in handle clone spawn_blocking: ThreadId(72)
current thread ID in thread: ThreadId(74)
current thread ID in spawn blocking from thread: ThreadId(72)
current thread ID in block on from runtime created in thread: ThreadId(74)
```
So what this tells us is that if we use `tokio::runtime::Builder::new_current_thread` to create a `tokio::runtime::Handle` in thread A, then clone that `Handle` to threads B, C, etc., then work sent to the cloned handles will all execute on thread A. We don't want this, because it would mean every cloned instance of `AwsSqsTaskQueue` would contend for a single thread.

So, to recap:
- we can't clone a `Runtime`
- we _can_ clone a `Handle`, but it won't do what we want to do
- we can't provide a custom `Clone` impl that instantiates a new `Runtime` because `clone()` can't fail

I'm not sure how to do better. We could have a field `Option<Runtime>` on `AwsSqsTaskQueue` and lazily initialize it from any method that needs the runtime, but honestly that seems like more trouble than it's worth unless we know that instantiating a `Runtime` is particularly expensive.

File: facilitator/src/aggregation.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jsha) [reply facilitator/src/aggregation.rs:116]: Yeah, I would treat `Watchdog` as a shared reference in all cases. As a side note, you don't necessarily have to clone the Watchdog to get a mutable instance; it might make more sense for your Watchdogs that actually mutate state (i.e. the ones that extend a Task's deadline) to use interior mutability for it.

File: facilitator/src/aggregation.rs [hunk 6]
```diff
@@ -94,61 +123,130 @@ impl<'a> BatchAggregator<'a> {
...
+                // Ideally, we would use the encryption_key_id in the ingestion
+                // packet to figure out which private key to use for decryption,
+                // but that field is optional[1]. Instead we try all the keys we
+                // have available until one works, meaning we have to create one
+                // Server per key.
+                // When there are no more jobs for this worker, we extract the
+                // accumulated results and send them to the queue to be reduced.
+                // Intuitively, we might expect to send results once for each
+                // dequeued job, but this approach lets us avoid re-allocating
+                // a Server for each iteration.
```
[the remark this answers is not in the document]
Comment (tgeoghegan) [reply facilitator/src/aggregation.rs:161]: You're right, I took another swing at it doing the more obvious and hypothetically slower thing and it's much nicer.

File: facilitator/src/aggregation.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (tgeoghegan) [reply facilitator/src/aggregation.rs:116]: The `TaskQueueWatchdog` already wraps its interesting fields in `Arc+Mutex` so I think that already achieves interior mutability. Put another way, I'm not sure what more would be achieved by sticking a `RefCell` somewhere between `Arc` and `Mutex`.

Sure enough, I found that I was able to just remove the `mut` from the `send_heartbeat(&mut self);` method on the `Watchdog` trait and everything Just Works, since `std::sync::Mutex::lock` takes `&self`.

**Review by jsha — APPROVED** [review]
Looks good to me! I left a comment about making TaskQueue do internal locking, but I think it's probably better to do that as a separate PR than to make this one even bigger. Relatedly: This PR seems to me to contain two major chunks: (1) changing the credential types to be more friendly to multithreading, and (2) multithreading the facilitator. I would tend to break out (1) into its own PR. I find that I often find interesting issues that way that I don't find when looking at a whole big PR. But up to you whether that is worth the tradeoff in time.

**Comment by tgeoghegan** [comment]
> Looks good to me! I left a comment about making TaskQueue do internal locking, but I think it's probably better to do that as a separate PR than to make this one even bigger. Relatedly: This PR seems to me to contain two major chunks: (1) changing the credential types to be more friendly to multithreading, and (2) multithreading the facilitator. I would tend to break out (1) into its own PR. I find that I often find interesting issues that way that I don't find when looking at a whole big PR. But up to you whether that is worth the tradeoff in time.

Thanks for the thorough review! You're right that this one sprawled just a bit too much and I should have done reviews the favor of chunking it up better. In the interest of avoiding further creep, I filed your wise suggestion of making `TaskQueue` impls threadsafe in #506.
